src/config/builder.rs

```diff
@@ -63,6 +63,53 @@ struct ConfigBuilderHash<'a> {
     secret: BTreeMap<&'a ComponentKey, &'a dyn SecretBackend>,
 }
 
+impl ConfigBuilderHash<'_> {
+    fn into_hash(self) -> String {
+        use sha2::{Digest, Sha256};
+
+        let value = self.to_sorted_string();
+        let output = Sha256::digest(value.as_bytes());
+
+        hex::encode(output)
+    }
+
+    /// Sort inner JSON values to maintain a consistent ordering. This
+    /// prevents non-deterministically serializable structures like HashMap from
+    /// affecting the resulting hash. As a consequence, ordering that does not
+    /// affect the actual semantics of a configuration is not considered when
+    /// calculating the hash.
+    fn to_sorted_string(self) -> String {
+        // Converting to Value prior to serializing to JSON string is sufficient
+        // to sort our underlying keys. This is because Value::Map is backed (by
+        // default) by BTreeMap which maintains an implicit key order.
+        // Serializing Value (based on Map) is thus deterministic versus
+        // serializing ConfigBuilderHash (based on potential HashMap) directly.
+        let value = serde_json::to_value(self)
```

Comment:
Are things like `HashSet` present/common in component configurations, etc.? We might want to explicitly sort `Value::Array`.

Reply:
Gotcha. I'm just worried that if anything like `HashSet` is present as the underlying store for a configuration option (or other), we'd face a similar non-deterministic serialization issue.

Reply:
Mmm, that's a good point.  I think we can probably just audit for `HashSet` and punt for now since I believe the configuration hashing strategy is changing shortly to just be a hash of the raw bytes of the configuration files. Otherwise, I think we would likely need some tests that would guard against us accidentally using a `HashSet` for configuration (or only sort those when creating the hash).

Reply:
Sounds good, that makes sense.